### PR TITLE
Create the charmhub refresher

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following updates the charmhub refresher to correctly handle
charmhub vs charmstore function calls. Essentially because we don't have
macaroons then we don't want to end up falling into the
AddCharmWithAuthorization workflow. By removing the special case we can
correctly reason about what the path the charmhub refresher will take.

In an ideal world we would use a better pattern for dealing with this.
Maybe resort to a stragtegy, but generally just embedding the
functionality into two different types also resolves this.

## QA steps

Unsure here?

### CODE REVIEW

File: cmd/juju/application/refresher/refresher.go [hunk 1]
```diff
@@ -116,15 +120,34 @@ func (d *factory) maybeReadLocal(charmAdder store.CharmAdder, charmRepo CharmRep
...
+
+func (d *factory) maybeCharmHub(charmAdder store.CharmAdder, charmResolver CharmResolver) func(RefresherConfig) (Refresher, error) {
+	return func(cfg RefresherConfig) (Refresher, error) {
+		return &charmHubRefresher{
+			baseRefresher: baseRefresher{
+				charmAdder:     charmAdder,
+				charmResolver:  charmResolver,
+				charmURL:       cfg.CharmURL,
+				charmRef:       cfg.CharmRef,
+				channel:        cfg.Channel,
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/refresher.go:146]: Can we pass an origin here instead?  To ensure the correct ID is run.  Or do you intend to change that in the follow ups?

File: cmd/juju/application/refresher/refresher.go [hunk 2]
```diff
@@ -280,3 +351,51 @@ type defaultCharmRepo struct{}
...
+
+// Refresh a given charm hub charm.
+// Bundles are not supported as there is no physical representation in Juju.
+func (r *charmHubRefresher) Refresh() (*CharmID, error) {
+	newURL, origin, err := r.ResolveCharm()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	curl, _, err := store.AddCharmFromURL(r.charmAdder, newURL, origin, r.force)
```
Comment (hmlanigan) [inline cmd/juju/application/refresher/refresher.go:353]: I'm wondering whether ignoring the returned origin is okay or not.  Where do you know that it has the correct revision and channel for the new charm to be install?

**Review by hmlanigan — APPROVED** [review]
A couple of questions, otherwise LGTM.

Verified that upgrading a charm for 2.8.3 and 2.9 controller with this client works.

File: cmd/juju/application/refresher/refresher.go (hunk 2 above)
Comment (SimonRichardson) [reply cmd/juju/application/refresher/refresher.go:353]: I have no idea, was going to ask you the exact same question

File: cmd/juju/application/refresher/refresher.go (hunk 1 above)
Comment (SimonRichardson) [reply cmd/juju/application/refresher/refresher.go:146]: Follow up

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
https://trello.com/c/G16mvD2d/2953-add-charmhub-refresh-scaffold

**Pull request merged**